src/py21cmfast/drivers/param_config.py

```diff
@@ -172,63 +209,124 @@ def is_compatible_with(self, other: InputParameters) -> bool:
 
-    def merge(self, other: InputParameters) -> InputParameters:
-        """Merge another InputParameters instance with this one, checking for compatibility."""
-        if not self.is_compatible_with(other):
-            raise ValueError(
-                f"Input parameters are not compatible. \n SELF {self} \n OTHER {other}"
-            )
-        return InputParameters(**{k: self[k] or other[k] for k in self.merge_keys()})
-
     @classmethod
-    def combine(cls, inputs: Sequence[InputParameters]) -> InputParameters:
-        """Combine multiple input parameter structs into one.
-
-        Parameters
-        ----------
-        inputs : list of :class:`InputParameters`
-            The input parameter structs to combine.
-
-        Returns
-        -------
-        :class:`InputParameters`
-            The combined input parameter struct.
+    def from_output_structs(cls, output_structs, **kwargs) -> InputParameters:
+        """Generate a new InputParameters instance given a list of OutputStructs.
+
+        In contrast to other construction methods, we do not accept overwriting of
+        sub-fields here, since it will no longer be compatible with the output structs.
+
+        All required fields not present in the `OutputStruct` objects need to be provided.
         """
-        # Create an empty instance of InputParameters
-        this = inputs[0]
+        # get matching fields in each output struct
+        fieldnames = [field.name for field in attrs.fields(cls) if field.eq]
+        input_params = {k: [] for k in fieldnames}
+        default_fields = []
+        for struct in output_structs:
+            if struct is None:
+                continue
+            for k in struct._inputs:
+                name = k.lstrip("_")
+                if name in fieldnames:
+                    input_params[name].append(getattr(struct, k, None))
+
+        # Now we have a list of [value | None,...] for each field,
+        for field, values in input_params.items():
+            # Append any provided structures
+            values.append(kwargs.pop(field, None))
+            # remove None and any duplicates
+            # NOTE: Types are not necessarily hashable so cannot use set
+            values = [
+                val
+                for i, val in enumerate(values)
+                if val is not None and val not in values[:i]
+            ]
+
+            # If we have multiple values, it means there is a clash
+            if len(values) > 1:
+                raise ValueError(
+                    f"InputParameters.from_output_struct got multiple values for {field}: {values}"
+                )
+            elif len(values) == 0:
+                if attrs.fields_dict(cls)[field].default:
+                    # If the parameter has a default, we want to remove them
+                    #    from the dict (after the loop)
+                    # TODO: this is messy, try to clean it up
+                    default_fields.append(field)
```

Comment:
why not just do `del input_params[field]` here, and then remove the line below that does the popping in the list comprension?

Reply:
We can't change the length of the dict inside it's own loop